rust/agents/relayer/src/msg/serial_submitter.rs

```diff
@@ -54,329 +55,243 @@ use super::{
+/// 2. Operations should be executed in in-order, i.e. if op_a was sent on
+/// source chain prior to op_b, and they're both destined for the same
+/// destination chain and are otherwise eligible, we should try to deliver op_a
+/// before op_b, all else equal. This is because we expect applications may
+/// prefer this even if they do not strictly rely on it for correctness.
 ///
 /// 3. Be [work-conserving](https://en.wikipedia.org/wiki/Work-conserving_scheduler) w.r.t.
 /// the single execution slot, i.e. so long as there is at least one message
 /// eligible for submission, we should be working on it within reason. This
 /// must be balanced with the cost of making RPCs that will almost certainly
 /// fail and potentially block new messages from being sent immediately.
-///
-/// TODO: Do we also want to await finality_blocks on source chain before
-///  attempting submission? Does this already happen?
 #[derive(Debug, new)]
-pub(crate) struct SerialSubmitter {
+pub struct SerialSubmitter {
+    /// Domain this submitter delivers to.
+    domain: HyperlaneDomain,
     /// Receiver for new messages to submit.
-    rx: mpsc::UnboundedReceiver<PendingMessage>,
-    /// Messages waiting for their turn to be dispatched. The SerialSubmitter
-    /// can only dispatch one message at a time, so this queue could grow.
-    #[new(default)]
-    run_queue: BinaryHeap<Reverse<PendingMessage>>,
-    /// Mailbox on the destination chain.
-    mailbox: CachingMailbox,
-    /// Used to construct the ISM metadata needed to verify a message.
-    metadata_builder: BaseMetadataBuilder,
-    /// Interface to agent rocks DB for e.g. writing delivery status upon
-    /// completion.
-    db: HyperlaneDB,
+    rx: mpsc::UnboundedReceiver<Box<DynPendingOperation>>,
     /// Metrics for serial submitter.
     metrics: SerialSubmitterMetrics,
-    /// Used to determine if messages have made sufficient gas payments.
-    gas_payment_enforcer: Arc<GasPaymentEnforcer>,
-    /// Hard limit on transaction gas when submitting a transaction.
-    transaction_gas_limit: Option<U256>,
 }
 
 impl SerialSubmitter {
-    pub fn spawn(mut self) -> Instrumented<JoinHandle<Result<()>>> {
-        tokio::spawn(async move { self.work_loop().await })
-            .instrument(info_span!("serial submitter work loop"))
+    pub fn spawn(self) -> Instrumented<JoinHandle<Result<()>>> {
+        let span = info_span!("SerialSubmitter", destination=%self.domain);
+        spawn(async move { self.run().await }).instrument(span)
     }
 
-    #[instrument(skip_all, fields(mbx=%self.mailbox.domain()))]
-    async fn work_loop(&mut self) -> Result<()> {
-        loop {
-            self.tick().await?;
-            sleep(Duration::from_millis(200)).await;
+    async fn run(self) -> Result<()> {
+        let Self {
+            domain,
+            metrics,
+            rx: rx_prepare,
+        } = self;
+        let prepare_queue: OpQueue = Default::default();
```

Comment:
nit: Consider comments explaining why prepare/confirm are queues, but submit is a channel